src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp

```diff
@@ -463,6 +465,30 @@ CTranslatorDXLToPlStmt::SetParamIds(Plan *plan)
 	plan->allParam = bitmapset;
 }
 
+//---------------------------------------------------------------------------
+//	@function:
+//		CTranslatorDXLToPlStmt::SetRewindPlanIds
+//
+//	@doc:
+//		Fills rewindPlanIDs bitmapset in PlannedStmt with plan_id of all
+//		SubPlans, except InitPlans.
+//---------------------------------------------------------------------------
+void
+CTranslatorDXLToPlStmt::SetRewindPlanIds(PlannedStmt *planned_stmt)
+{
+	Bitmapset *bitmapset = NULL;
+	List *subplan_exprs_list = m_dxl_to_plstmt_context->GetSubPlanExprsList();
+	ListCell *lc = NULL;
+
+	ForEach(lc, subplan_exprs_list)
+	{
+		SubPlan *subplan = (SubPlan *) lfirst(lc);
+		if (!subplan->is_initplan)
+			bitmapset = gpdb::BmsAddMember(bitmapset, subplan->plan_id);
+	}
+
+	planned_stmt->rewindPlanIDs = bitmapset;
```

Comment:
Did you consider `CTranslatorDXLToPlStmt::SetInitPlanVariables` as a place for this logic?

Reply:
Yes, I did. This function suits well for our purposes. However, how legal is to change the function's 1) naming, 2) specification and comments.

We can make the ` SetRewindPlanIds` function traverse the whole tree to find InitPlans as  in `SetInitPlanVariables`. But I think it's more expensive, than supporting the state list ` m_subplan_exprs_list`.

Reply:
Thus, what is better:  change `SetInitPlanVariables` and fill the rewindPlansIDs during tree traversal, or add  `SetRewindPlanIds` function? And if we add a separate function is it better to maintain another list of SubPlan's or implement another tree traversal?

Reply:
In my opinion, the presence of both approaches together looks redundant. If you want to implement a separate list of subplans, you need to rework `SetInitPlanVariables` too. But, it's more simple to change function name (e.g. to `SetSubPlanVariables`) and description and collect all non-init-plans to bitmap here.